# Load Dataset Diabete.csv

In [1]:
# downlad dataset bike_buyer.csv
!wget --no-check-certificate 'https://docs.google.com/uc?export=download&id=12ffreGLdc-g9nIxQ1JGKk2hVB9-NLoTi' -O diabete.csv

--2023-09-21 13:33:36--  https://docs.google.com/uc?export=download&id=12ffreGLdc-g9nIxQ1JGKk2hVB9-NLoTi
Resolving docs.google.com (docs.google.com)... 172.253.63.102, 172.253.63.138, 172.253.63.139, ...
Connecting to docs.google.com (docs.google.com)|172.253.63.102|:443... connected.
HTTP request sent, awaiting response... 303 See Other
Location: https://doc-00-28-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/3p2u94g937knepki3knib5ckcm567e2n/1695303150000/05830297694632628921/*/12ffreGLdc-g9nIxQ1JGKk2hVB9-NLoTi?e=download&uuid=4ff035bd-8b3a-49f3-8893-4cfe2f19b422 [following]
--2023-09-21 13:33:36--  https://doc-00-28-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/3p2u94g937knepki3knib5ckcm567e2n/1695303150000/05830297694632628921/*/12ffreGLdc-g9nIxQ1JGKk2hVB9-NLoTi?e=download&uuid=4ff035bd-8b3a-49f3-8893-4cfe2f19b422
Resolving doc-00-28-docs.googleusercontent.com (doc-00-28-docs.googleusercontent.com)... 142.251.16.132, 2607:f8b0:

In [5]:
import pandas as pd
df = pd.read_csv('diabete.csv')
print(df.head(5))

   PatientID  NumberOfPregnant  Glucose  BloodPressure  SkinThickness  \
0          1                 6      148             72             35   
1          2                 1       85             66             29   
2          3                 8      183             64              0   
3          4                 1       89             66             23   
4          5                 0      137             40             35   

   Insulin   BMI  DiabetesPedigree  Age  HasDiabetes  
0        0  33.6             0.627   50            1  
1        0  26.6             0.351   31            0  
2        0  23.3             0.672   32            1  
3       94  28.1             0.167   21            0  
4      168  43.1             2.288   33            1  


In [6]:
df.drop(['PatientID'], axis = 1, inplace = True)

In [7]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   NumberOfPregnant  768 non-null    int64  
 1   Glucose           768 non-null    int64  
 2   BloodPressure     768 non-null    int64  
 3   SkinThickness     768 non-null    int64  
 4   Insulin           768 non-null    int64  
 5   BMI               768 non-null    float64
 6   DiabetesPedigree  768 non-null    float64
 7   Age               768 non-null    int64  
 8   HasDiabetes       768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB
None


In [8]:
# check duplicate row
print(df.duplicated().sum())

0


In [9]:
X = df.drop('HasDiabetes', axis = 1)
y = df['HasDiabetes']

In [10]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=11)

# TO_DO: Exercise - Load dataset Iris

In [1]:
!wget --no-check-certificate 'https://docs.google.com/uc?export=download&id=1oUN5YCdQisXIYQIGioSIe75KSsJzdxPl' -O iris.csv

--2023-09-22 18:31:20--  https://docs.google.com/uc?export=download&id=1oUN5YCdQisXIYQIGioSIe75KSsJzdxPl
Resolving docs.google.com (docs.google.com)... 74.125.26.113, 74.125.26.100, 74.125.26.102, ...
Connecting to docs.google.com (docs.google.com)|74.125.26.113|:443... connected.
HTTP request sent, awaiting response... 303 See Other
Location: https://doc-14-28-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/qgo4h5qaf35p6keo4epsuudeq78hnsee/1695407475000/05830297694632628921/*/1oUN5YCdQisXIYQIGioSIe75KSsJzdxPl?e=download&uuid=6cbe9bec-3a20-4197-8bbf-b727496a224a [following]
--2023-09-22 18:31:20--  https://doc-14-28-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/qgo4h5qaf35p6keo4epsuudeq78hnsee/1695407475000/05830297694632628921/*/1oUN5YCdQisXIYQIGioSIe75KSsJzdxPl?e=download&uuid=6cbe9bec-3a20-4197-8bbf-b727496a224a
Resolving doc-14-28-docs.googleusercontent.com (doc-14-28-docs.googleusercontent.com)... 173.194.218.132, 2607:f8b0:400

In [2]:
import pandas as pd
iris_df = pd.read_csv('iris.csv')
print(iris_df)

      Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm  \
0      1            5.1           3.5            1.4           0.2   
1      2            4.9           3.0            1.4           0.2   
2      3            4.7           3.2            1.3           0.2   
3      4            4.6           3.1            1.5           0.2   
4      5            5.0           3.6            1.4           0.2   
..   ...            ...           ...            ...           ...   
145  146            6.7           3.0            5.2           2.3   
146  147            6.3           2.5            5.0           1.9   
147  148            6.5           3.0            5.2           2.0   
148  149            6.2           3.4            5.4           2.3   
149  150            5.9           3.0            5.1           1.8   

            Species  
0       Iris-setosa  
1       Iris-setosa  
2       Iris-setosa  
3       Iris-setosa  
4       Iris-setosa  
..              ...  
145  

In [3]:
# Check duplicate row
print(iris_df.duplicated().sum())

0


In [4]:
# Handle duplicates
iris_df.drop_duplicates(inplace = True)

In [5]:
iris_df = iris_df.drop("Id", axis = 1)
iris_X = iris_df.drop("Species", axis = 1)
iris_y = iris_df["Species"]

In [6]:
d = {'Iris-setosa': 0, 'Iris-versicolor': 1, 'Iris-virginica': 2}
iris_y = iris_y.map(d)

In [8]:
from sklearn.model_selection import train_test_split
iris_X_train, iris_X_test, iris_y_train, iris_y_test = train_test_split(iris_X, iris_y, test_size=0.2, random_state=11)

#1.1Naive Bayes

## Naive Bayes - Cross Validation

In [39]:
from sklearn.model_selection import cross_val_score
from sklearn.naive_bayes import GaussianNB

NB_model = GaussianNB()
cvs = cross_val_score(NB_model, X, y, cv=5)
print('cross val scores', cvs.round(3))
print('mean = ', cvs.mean().round(4) * 100)
print('std = ', (cvs.std() * 100).round(3))

cross val scores [0.753 0.727 0.747 0.784 0.745]
mean =  75.13
std =  1.86


## Naive Bayes - Train/Test Split

In [40]:
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
from sklearn.naive_bayes import GaussianNB

NB_model = GaussianNB()
NB_model = NB_model.fit(X_train, y_train)

y_pred = NB_model.predict(X_test)
print('Accuracy Score: ', accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))
cm = confusion_matrix(y_test, y_pred)
print(cm)

Accuracy Score:  0.7402597402597403
              precision    recall  f1-score   support

           0       0.78      0.84      0.81       100
           1       0.65      0.56      0.60        54

    accuracy                           0.74       154
   macro avg       0.71      0.70      0.70       154
weighted avg       0.73      0.74      0.73       154

[[84 16]
 [24 30]]


show the data that predict false

In [41]:
import pandas as pd
dy = pd.DataFrame()
dy['y_test'] = y_test
dy['y_pred'] = y_pred
dy['result'] = y_test == y_pred

dy[dy.result == False]

y_test  y_pred  result
153       0       1   False
66        1       0   False
143       1       0   False
148       0       1   False
533       0       1   False
25        1       0   False
295       0       1   False
510       1       0   False
569       1       0   False
395       0       1   False
701       1       0   False
593       0       1   False
664       1       0   False
144       0       1   False
356       1       0   False
349       1       0   False
223       0       1   False
129       1       0   False
128       1       0   False
309       1       0   False
164       1       0   False
766       1       0   False
214       1       0   False
293       1       0   False
397       1       0   False
495       0       1   False
95        0       1   False
645       0       1   False
6         1       0   False
499       0       1   False
188       1       0   False
361       0       1   False
709       1       0   False
696       1       0   False
247       0       1   False
70        1       0   False
402       1       0   False
453       0       1   False
57        0       1   False
93        1       0   False

In [42]:
print(df.loc[[200]])

     NumberOfPregnant  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
200                 0      113             80             16        0  31.0   

     DiabetesPedigree  Age  HasDiabetes  
200             0.874   21            0  


#1.2Naive Bay (Iris_Dataset)

##Naive Bay - Cross Validataion (Iris_Dataset)

In [9]:
# TO_DO
from sklearn.model_selection import cross_val_score
from sklearn.naive_bayes import GaussianNB

naive_bays_classifier = GaussianNB()
cvs = cross_val_score(naive_bays_classifier, iris_X, iris_y, cv=5)
print('cross val scores', cvs.round(3))
print('mean = ', cvs.mean().round(4) * 100)
print('std = ', (cvs.std() * 100).round(3))

cross val scores [0.933 0.967 0.933 0.933 1.   ]
mean =  95.33
std =  2.667


## Analysis:

1.Cross-Validation Scores:
These scores [0.933 0.967 0.933 0.933 1.] signify the accuracy of the model for each of the five folds. The scores are significantly high, indicating strong performance of the model on each fold.

2.Mean Accuracy:
The mean accuracy 95.33 suggests that the model is, on average, able to correctly predict the class labels 95.33% of the time. This indicates a high level of performance of the model on the dataset.

3.Standard Deviation of Accuracy:
The standard deviation std = 2.667 is a measure of the amount of variation or dispersion in the accuracy scores. A lower value of standard deviation is usually desired, indicating consistent performance across different subsets of data. However, in this case, the standard deviation is relatively higher compared to the mean, signifying some level of inconsistency in the model's performance across different folds.

##Naive Bay - Train/Test Split (Iris_Dataset)

In [10]:
# TO_DO
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
from sklearn.naive_bayes import GaussianNB

# Training the Naive Bayes classifier
naive_bays_classifier2 = GaussianNB()
naive_bays_classifier2.fit(iris_X_train, iris_y_train)

# Predicting using the trained model
y_pred = naive_bays_classifier2.predict(iris_X_test)

# Evaluating the predictions
# Use the iris_y_test for evaluation
print('Accuracy Score: ', accuracy_score(iris_y_test, y_pred))
print(classification_report(iris_y_test, y_pred))
cm = confusion_matrix(iris_y_test, y_pred)
print(cm)

Accuracy Score:  0.9333333333333333
              precision    recall  f1-score   support

           0       1.00      1.00      1.00         9
           1       0.83      1.00      0.91        10
           2       1.00      0.82      0.90        11

    accuracy                           0.93        30
   macro avg       0.94      0.94      0.94        30
weighted avg       0.94      0.93      0.93        30

[[ 9  0  0]
 [ 0 10  0]
 [ 0  2  9]]


#2.1k-Nearest Neighbor

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

knn_model = KNeighborsClassifier(n_neighbors=3, metric='euclidean')
knn_model.fit(X_train,y_train)

y_pred = knn_model.predict(X_test)

print('Accuracy Score: ', accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))
cm = confusion_matrix(y_test, y_pred)
print(cm)


Accuracy Score:  0.7012987012987013
              precision    recall  f1-score   support

           0       0.74      0.83      0.78       100
           1       0.60      0.46      0.52        54

    accuracy                           0.70       154
   macro avg       0.67      0.65      0.65       154
weighted avg       0.69      0.70      0.69       154

[[83 17]
 [29 25]]


#2.2k-Nearest Neighbor (Iris_Dataset)

In [11]:
# TO_DO
from sklearn.neighbors import KNeighborsClassifier

knn_model = KNeighborsClassifier(n_neighbors=3, metric='euclidean')
knn_model.fit(iris_X_train, iris_y_train)

y_pred = knn_model.predict(iris_X_test)

print('Accuracy Score: ', accuracy_score(iris_y_test, y_pred))
print(classification_report(iris_y_test, y_pred))
cm = confusion_matrix(iris_y_test, y_pred)
print(cm)

Accuracy Score:  0.9333333333333333
              precision    recall  f1-score   support

           0       1.00      1.00      1.00         9
           1       0.83      1.00      0.91        10
           2       1.00      0.82      0.90        11

    accuracy                           0.93        30
   macro avg       0.94      0.94      0.94        30
weighted avg       0.94      0.93      0.93        30

[[ 9  0  0]
 [ 0 10  0]
 [ 0  2  9]]


#3.1Logistic Regression

In [ ]:
from sklearn.linear_model import LogisticRegression

logreg_model = LogisticRegression()
logreg_model.fit(X_train, y_train)

y_pred = logreg_model.predict(X_test)

print('Accuracy Score: ', accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))
cm = confusion_matrix(y_test, y_pred)
print(cm)

Accuracy Score:  0.7402597402597403
              precision    recall  f1-score   support

           0       0.75      0.89      0.82       100
           1       0.69      0.46      0.56        54

    accuracy                           0.74       154
   macro avg       0.72      0.68      0.69       154
weighted avg       0.73      0.74      0.73       154

[[89 11]
 [29 25]]


/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


#4.1SVM

In [ ]:
from sklearn.svm import SVC

svm_model = SVC(kernel='linear')
svm_model.fit(X_train, y_train)

y_pred = svm_model .predict(X_test)

print('Accuracy Score: ', accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))
cm = confusion_matrix(y_test, y_pred)
print(cm)

Accuracy Score:  0.7402597402597403
              precision    recall  f1-score   support

           0       0.75      0.89      0.82       100
           1       0.69      0.46      0.56        54

    accuracy                           0.74       154
   macro avg       0.72      0.68      0.69       154
weighted avg       0.73      0.74      0.73       154

[[89 11]
 [29 25]]


#4.2SVM (Iris_Dataset)

In [12]:
# TO_DO
from sklearn.svm import SVC

svm_model = SVC(kernel='linear')
svm_model.fit(iris_X_train, iris_y_train)

y_pred = svm_model .predict(iris_X_test)

print('Accuracy Score: ', accuracy_score(iris_y_test, y_pred))
print(classification_report(iris_y_test, y_pred))
cm = confusion_matrix(iris_y_test, y_pred)
print(cm)


Accuracy Score:  1.0
              precision    recall  f1-score   support

           0       1.00      1.00      1.00         9
           1       1.00      1.00      1.00        10
           2       1.00      1.00      1.00        11

    accuracy                           1.00        30
   macro avg       1.00      1.00      1.00        30
weighted avg       1.00      1.00      1.00        30

[[ 9  0  0]
 [ 0 10  0]
 [ 0  0 11]]


#5.1Random Forest

In [ ]:
from sklearn.ensemble import RandomForestClassifier

RF_model = RandomForestClassifier(n_estimators=80) #The number of trees
RF_model.fit(X_train, y_train)

y_pred = RF_model.predict(X_test)

print('Accuracy Score: ', accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))
cm = confusion_matrix(y_test, y_pred)
print(cm)

Accuracy Score:  0.7337662337662337
              precision    recall  f1-score   support

           0       0.76      0.86      0.81       100
           1       0.66      0.50      0.57        54

    accuracy                           0.73       154
   macro avg       0.71      0.68      0.69       154
weighted avg       0.73      0.73      0.72       154

[[86 14]
 [27 27]]


#5.2Random Forest (Iris_Dataset)

In [13]:
# TO_DO
from sklearn.ensemble import RandomForestClassifier

RF_model = RandomForestClassifier(n_estimators=80) #The number of trees
RF_model.fit(iris_X_train, iris_y_train)

y_pred = RF_model.predict(iris_X_test)

print('Accuracy Score: ', accuracy_score(iris_y_test, y_pred))
print(classification_report(iris_y_test, y_pred))
cm = confusion_matrix(iris_y_test, y_pred)
print(cm)

Accuracy Score:  0.9333333333333333
              precision    recall  f1-score   support

           0       1.00      1.00      1.00         9
           1       0.83      1.00      0.91        10
           2       1.00      0.82      0.90        11

    accuracy                           0.93        30
   macro avg       0.94      0.94      0.94        30
weighted avg       0.94      0.93      0.93        30

[[ 9  0  0]
 [ 0 10  0]
 [ 0  2  9]]


#6.1Artificial Neural Network

In [ ]:
from sklearn.neural_network import MLPClassifier

ANN_model = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(5, 2), random_state=1)

ANN_model.fit(X, y)

y_pred = ANN_model.predict(X_test)

print('Accuracy Score: ', accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))
cm = confusion_matrix(y_test, y_pred)
print(cm)

Accuracy Score:  0.6558441558441559
              precision    recall  f1-score   support

           0       0.68      0.90      0.77       100
           1       0.52      0.20      0.29        54

    accuracy                           0.66       154
   macro avg       0.60      0.55      0.53       154
weighted avg       0.62      0.66      0.60       154

[[90 10]
 [43 11]]


#6.2Neural Network (Iris_Dataset)

In [14]:
# TO_DO
from sklearn.neural_network import MLPClassifier

ANN_model = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(5, 2), random_state=1)

ANN_model.fit(iris_X_train, iris_y_train)

y_pred = ANN_model.predict(iris_X_test)

print('Accuracy Score: ', accuracy_score(iris_y_test, y_pred))
print(classification_report(iris_y_test, y_pred))
cm = confusion_matrix(iris_y_test, y_pred)
print(cm)


Accuracy Score:  0.3
              precision    recall  f1-score   support

           0       0.30      1.00      0.46         9
           1       0.00      0.00      0.00        10
           2       0.00      0.00      0.00        11

    accuracy                           0.30        30
   macro avg       0.10      0.33      0.15        30
weighted avg       0.09      0.30      0.14        30

[[ 9  0  0]
 [10  0  0]
 [11  0  0]]


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
